va/tlsalpn.go

```diff
@@ -160,26 +156,21 @@ func (va *ValidationAuthorityImpl) getChallengeCert(
 	return certs[0], &cs, nil
 }
 
-// tlsDial does the equivalent of tls.Dial, but obeying a context. Once
-// tls.DialContextWithDialer is available, switch to that.
+// tlsDial does the equivalent of tls.Dial, but obeying a context for both the
+// dial and handshake operations.
 func (va *ValidationAuthorityImpl) tlsDial(ctx context.Context, hostPort string, config *tls.Config) (*tls.Conn, error) {
 	ctx, cancel := context.WithTimeout(ctx, va.singleDialTimeout)
 	defer cancel()
 	dialer := &net.Dialer{}
 	netConn, err := dialer.DialContext(ctx, "tcp", hostPort)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("during dial: %w", err)
 	}
-	deadline, ok := ctx.Deadline()
-	if !ok {
-		va.log.AuditErr("tlsDial was called without a deadline")
-		return nil, fmt.Errorf("tlsDial was called without a deadline")
-	}
-	_ = netConn.SetDeadline(deadline)
+
 	conn := tls.Client(netConn, config)
-	err = conn.Handshake()
+	err = conn.HandshakeContext(ctx)
```

Comment:
Ah very exciting to see HandshakeContext! I guess DialContextWithDialer isn't gonna happen. 😢

Reply:
Yeah, that was my conclusion too :(

Reply:
Looking at the docs again, I think we can get what we want with [`tls.Dialer`](https://pkg.go.dev/crypto/tls#Dialer), containing our custom dialer. We would then call `tlsDialer.DialContext`, which would perform both the TCP connect and the handshake.

Reply:
Good catch, this appears to work! PTAL